pkg/boot/fit/vfit.go

```diff
@@ -0,0 +1,211 @@
+	fmt.Stringer
+	// Warning: If the signature does not exist or does not match the keyring,
+	// both the file and a signature error will be returned.
+	Verify([]byte, openpgp.KeyRing) (*bytes.Reader, error)
+}
+
+type PGPSignature struct {
+	name  string // Name of signature Node
+	value []byte
+	// Optional description fields
+	signer string
+	hint   string
+}
+
+type RSASignature struct {
+	name  string // Name of signature Node
+	hash  crypto.Hash
+	value []byte
+	// Optional description fields
+	signer string
+	hint   string
+}
+
+func (s PGPSignature) String() string {
+	return fmt.Sprintf("PGP Signature - name: %s, signer: '%s', hint: '%s'", s.name, s.signer, s.hint)
+}
+
+// Warning: If the signature does not exist or does not match the keyring,
+// both the file and a signature error will be returned.
+func (s PGPSignature) Verify(b []byte, ring openpgp.KeyRing) (*bytes.Reader, error) {
+	r := bytes.NewReader(b)
+	if signer, err := openpgp.CheckDetachedSignature(ring, bytes.NewReader(b), bytes.NewReader(s.value)); err != nil {
+		return r, vfile.ErrUnsigned{Path: s.name, Err: err}
+	} else if signer == nil {
+		return r, vfile.ErrUnsigned{Path: s.name, Err: vfile.ErrWrongSigner{ring}}
+	}
+	return r, nil
+}
+
+// Warning: If the signature does not exist or does not match the keyring,
+// both the file and a signature error will be returned.
+func (s RSASignature) Verify(b []byte, ring openpgp.KeyRing) (*bytes.Reader, error) {
+	r := bytes.NewReader(b)
+	keys, err := vfile.GetRSAKeysFromRing(ring)
+	if err != nil {
+		return r, err
+	}
+
+	hashed, err := vfile.CalculateHash(bytes.NewReader(b), s.hash.New())
+	for _, key := range keys {
+		if err = rsa.VerifyPKCS1v15(key, s.hash, hashed, s.value); err == nil {
+			return r, nil
+		}
+	}
+	return r, vfile.ErrUnsigned{Err: vfile.ErrWrongSigner{ring}}
+}
+
+func (s RSASignature) String() string {
+	return fmt.Sprintf("RSA Signature - name: %s, signer: '%s', hint: '%s', hash: '%s'", s.name, s.signer, s.hint, s.hash)
+}
+
+// parseHash cleans and maps a string into a crytpo.Hash.
+// Expected format: "sha256,rsa4096" or "sha1"
+func parseHash(algo string) (crypto.Hash, error) {
+	cleaned := strings.TrimFunc(algo, func(r rune) bool {
+		return !unicode.IsGraphic(r)
+	})
+	algoSplit := strings.Split(cleaned, ",")
+	if len(algoSplit) == 0 {
+		return 0, fmt.Errorf("Unrecognized hash algo: '%s'", cleaned)
+	}
+	if alg, ok := algs[strings.ToUpper(algoSplit[0])]; ok {
+		return alg, nil
+	}
+	return 0, fmt.Errorf("Unrecognized hash algo: '%s'", cleaned)
+}
+
+// parseSignatures parses dt.Node to RSASignatures
+// Nodes with missing required properties or invalid hash functions are skipped.
+// An error is returned if no valid signatures are found
+func parseSignatures(n ...*dt.Node) ([]Signature, error) {
+	var sigs []Signature
+	for _, node := range n {
+		v, ok := node.LookProperty("value")
+		if !ok {
+			fmt.Printf("Skipping signature %s: missing value node", node.Name)
```

Comment:
Is it intended for this library to printf as it works, or is this possibly leftover from testing?

Reply:
Hmm.. I'll ask the wider audience what is expected of uroot pkg libraries. I personally would prefer to leave the print. I don't see a reasonable way to return this data to the caller. Verbose parameters/globals and I think this would be important enough to print in most cases anyways.

Reply:
Sounds good. Just wanted to make sure this was intentional.